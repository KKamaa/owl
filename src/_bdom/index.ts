import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
