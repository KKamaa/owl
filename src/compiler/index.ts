import { setupBlockDom } from "./helpers";
import { QWebCompiler, TemplateFunction } from "./qweb_compiler";
export { Template, TemplateFunction } from "./qweb_compiler";

setupBlockDom();

export function compileTemplate(template: string, name?: string): TemplateFunction {
  const compiler = new QWebCompiler(template, name);
  return compiler.compile();
}
