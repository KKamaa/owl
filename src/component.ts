import type { BNode } from "./b_node";

export class Component {
  static template: string = "";

  props: any;
  env: any;
  __owl__: BNode;

  constructor(props: any, env: any, node: BNode) {
    this.props = props;
    this.env = env;
    this.__owl__ = node;
  }
  get el(): HTMLElement | null {
    const node = this.__owl__;
    return node.bdom ? (node.bdom.el as any) : null;
  }

  setup() {}

  render(): Promise<void> {
    return this.__owl__.render();
  }
}
