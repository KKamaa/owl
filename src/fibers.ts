import type { Block } from "./bdom";
import { STATUS } from "./status";
import { OwlNode } from "./owl_node";

export function makeChildFiber(node: OwlNode, parent: Fiber): Fiber {
  // todo: remove this
  if ((window as any).debug) {
    console.log("childfiber for " + node.component.constructor.name);
  }
  let current = node.fiber;
  if (current) {
    // current is necessarily a rootfiber here
    let root = parent.root;
    cancelFibers(current.children);
    current.children = [];
    current.parent = parent;
    root.counter++;
    return current;
  }
  let result = new Fiber(node, parent);
  return result;
}

export function makeRootFiber(node: OwlNode): Fiber {
  // todo: remove this
  if ((window as any).debug) {
    console.log("rootfiber for " + node.component.constructor.name);
  }
  let current = node.fiber;
  if (current) {
    let root = current.root;
    root.counter -= cancelFibers(current.children);
    current.children = [];
    root.counter++;
    current.bdom = null;
    return current;
  }
  let result = new RootFiber(node);
  return result;
}

export function makeMountFiber(node: OwlNode, target: HTMLElement): MountFiber {
  if ((window as any).debug) {
    console.log("mountfiber for " + node.component.constructor.name);
  }
  let result = new MountFiber(node, target);
  return result;
}

/**
 * @returns number of not-yet rendered fibers cancelled
 */
function cancelFibers(fibers: Fiber[]): number {
  let result = 0;
  for (let fiber of fibers) {
    fiber.node.fiber = null;
    if (!fiber.bdom) {
      result++;
    }
    result += cancelFibers(fiber.children);
  }
  return result;
}

export class Fiber {
  node: OwlNode;
  bdom: Block | null = null;
  root: RootFiber;
  parent: Fiber | null;
  children: Fiber[] = [];

  constructor(node: OwlNode, parent: Fiber | null) {
    this.node = node;
    node.fiber = this;
    this.parent = parent;
    if (parent) {
      const root = parent.root;
      root.counter++;
      this.root = root;
      parent.children.push(this);
    } else {
      this.root = this as any;
    }
  }
}

export class RootFiber extends Fiber {
  counter: number = 1;
  error: Error | null = null;
  resolve: any;
  promise: any;
  reject: any;
  toPatch: Fiber[];

  constructor(node: OwlNode) {
    super(node, null);
    this.counter = 1;
    this.error = null;
    this.toPatch = [this];

    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }

  complete() {
    const node = this.node;
    // if (node.fiber !== fiber) {
    //   return;
    // }
    for (let fiber of this.toPatch) {
      let node = fiber.node;
      // because of the asynchronous nature of the rendering, some parts of the
      // UI may have been rendered, then deleted in a followup rendering, and we
      // do not want to call onBeforePatch in that case.
      if (node.fiber === fiber) {
        node.callBeforePatch();
      }
    }
    const mountedNodes: any[] = [];
    const patchedNodes: any[] = [node];
    node.bdom!.patch(this.bdom!, mountedNodes, patchedNodes);
    this.finalize(mountedNodes, patchedNodes);
    node.fiber = null;
  }

  finalize(mounted: OwlNode[], patched: OwlNode[]) {
    let current;
    while ((current = mounted.pop())) {
      for (let cb of current.mounted) {
        cb();
      }
    }
    while ((current = patched.pop())) {
      for (let cb of current.patched) {
        cb();
      }
    }
  }
}

export class MountFiber extends RootFiber {
  target: HTMLElement;

  constructor(node: OwlNode, target: HTMLElement) {
    super(node);
    this.target = target;
    this.toPatch = [];
  }
  complete() {
    const node = this.node;
    node.bdom = this.bdom;
    const mountedNodes: any[] = [node];
    const patchedNodes: any[] = [];
    node.bdom!.mount(this.target, mountedNodes, patchedNodes);
    this.finalize(mountedNodes, patchedNodes);
    node.status = STATUS.MOUNTED;
    node.fiber = null;
  }
}
