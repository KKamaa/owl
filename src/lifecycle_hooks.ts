import { getCurrent } from "./core/owl_node";

// -----------------------------------------------------------------------------
//  hooks
// -----------------------------------------------------------------------------

export function onWillStart(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.willStart.push(fn);
}

export function onBeforeUnmount(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.beforeUnmount.push(fn);
}

export function onMounted(fn: () => void) {
  const node = getCurrent()!;
  node.mounted.push(fn);
}

export function onBeforePatch(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.beforePatch.push(fn);
}

export function onPatched(fn: () => void) {
  const node = getCurrent()!;
  node.patched.push(fn);
}
