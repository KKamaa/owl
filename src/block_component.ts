import { Block } from "./bdom";
import { makeChildFiber, __internal__destroyed } from "./fibers";
import { OwlNode } from "./owl_node";
import { STATUS } from "./status";

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

export class BComponent extends Block {
  node: OwlNode;
  parentClass?: any;
  classTarget?: HTMLElement;

  /**
   *
   * @param name
   * @param props
   * @param key
   * @param owner the component in which the component was defined
   * @param parent the actual parent (may be different in case of slots)
   */
  constructor(name: string, props: any, key: string, owner: any, parent: any) {
    super();
    const parentNode: OwlNode = parent.__owl__;
    let node: OwlNode | undefined = parentNode.children[key];
    if (node && node.status < STATUS.MOUNTED) {
      node.destroy();
      // delete parentNode.children[key];
      node = undefined;
    }
    const parentFiber = parentNode.fiber!;
    if (node) {
      // update
      const fiber = makeChildFiber(node, parentFiber);
      if (node.willPatch.length) {
        parentFiber.root.willPatch.push(fiber);
      }
      if (node.patched.length) {
        parentFiber.root.patched.push(fiber);
      }
      node.updateAndRender(props, fiber);
    } else {
      // new component
      const components = owner.constructor.components;
      const C = components[name];
      node = new OwlNode(parentNode.app, C, props);
      parentNode.children[key] = node;
      const fiber = makeChildFiber(node, parentNode.fiber!);
      if (node.mounted.length) {
        parentFiber.root.mounted.push(fiber);
      }
      node.initiateRender(fiber);
    }
    this.node = node;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.node.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const node = this.node;
    const bdom = node.fiber!.bdom!;
    node.bdom = bdom;
    bdom.mountBefore(anchor);
    if (this.parentClass) {
      this.parentClass = this.parentClass.trim().split(/\s+/);
      const el = this.firstChildNode();
      if (el instanceof HTMLElement) {
        this.addClass(el);
      }
    }
    node.status = STATUS.MOUNTED;
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  moveBefore(anchor: ChildNode) {
    this.node.bdom!.moveBefore(anchor);
  }

  addClass(el: HTMLElement) {
    this.classTarget = el;
    for (let cl of this.parentClass) {
      el.classList.add(cl);
    }
  }

  removeClass(el: HTMLElement) {
    for (let cl of this.parentClass) {
      el.classList.remove(cl);
    }
  }

  patch() {
    const node = this.node;

    node.bdom!.patch(node!.fiber!.bdom!);
    if (this.parentClass) {
      const el = this.firstChildNode();
      if (el !== this.classTarget) {
        if (el && this.classTarget) {
          this.removeClass(this.classTarget);
          this.addClass(el as any);
        } else if (el) {
          this.addClass(el as any);
        } else {
          this.removeClass(this.classTarget!);
          this.classTarget = undefined;
        }
      }
    }
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  beforeRemove() {
    visitRemovedNodes(this.node);
  }

  remove() {
    const bdom = this.node.bdom!;
    bdom.remove();
  }
}

function visitRemovedNodes(node: OwlNode) {
  if (node.status === STATUS.MOUNTED) {
    const component = node.component;
    for (let cb of node.willUnmount) {
      cb.call(component);
    }
  }
  for (let child of Object.values(node.children)) {
    visitRemovedNodes(child);
  }
  node.status = STATUS.DESTROYED;
  if (node.destroyed.length) {
    __internal__destroyed.push(node);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, owner: any, parent: any) {
    super(name, props, key, owner, parent);
    this.handlers = new Array(handlers);
  }
  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }
  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.node.component.el!;
      el.addEventListener(eventType, (ev: Event) => {
        const info = this.handlers![i];
        const [, callback] = info;
        callback(ev);
      });
    }
  }
}
