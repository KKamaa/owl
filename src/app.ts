import { Blocks as BaseBlocks } from "./bdom";
import { compileTemplate, Template } from "./compiler/index";
import { globalTemplates } from "./tags";
import { BComponent, BComponentH } from "./block_component";
import type { Component } from "./component";
import { OwlNode } from "./owl_node";
import { Scheduler } from "./scheduler";
import { UTILS } from "./template_utils";

const Blocks = {
  ...BaseBlocks,
  BComponent,
  BComponentH,
};

// -----------------------------------------------------------------------------
//  TemplateSet
// -----------------------------------------------------------------------------

export class TemplateSet {
  rawTemplates: { [name: string]: string } = Object.create(globalTemplates);
  templates: { [name: string]: Template } = {};
  utils: typeof UTILS;

  constructor() {
    const call = (subTemplate: string, ctx: any, parent: any) => {
      const template = this.getTemplate(subTemplate);
      return template(ctx, parent);
    };

    this.utils = Object.assign({}, UTILS, { call });
  }

  addTemplate(name: string, template: string, options: { allowDuplicate?: boolean } = {}) {
    if (name in this.rawTemplates && !options.allowDuplicate) {
      throw new Error(`Template ${name} already defined`);
    }
    this.rawTemplates[name] = template;
  }

  getTemplate(name: string): Template {
    if (!(name in this.templates)) {
      const rawTemplate = this.rawTemplates[name];
      if (rawTemplate === undefined) {
        throw new Error(`Missing template: "${name}"`);
      }
      const templateFn = compileTemplate(rawTemplate, name);
      const template = templateFn(Blocks, this.utils);
      this.templates[name] = template;
    }
    return this.templates[name];
  }
}

export class App<T extends typeof Component = any> extends TemplateSet {
  Root: T;
  props: any;
  env: any = {};
  scheduler = new Scheduler(window.requestAnimationFrame.bind(window));
  root: OwlNode | null = null;
  dev: boolean = true;
  
  constructor(Root: T, props?: any) {
    super();
    this.Root = Root;
    this.props = props;
  }

  configure(params: any) {
    if (params.env) {
      this.env = params.env;
    }
  }

  mount(target: HTMLElement): Promise<InstanceType<T>> {
    if (!(target instanceof HTMLElement)) {
      throw new Error("Cannot mount component: the target is not a valid DOM element");
    }
    if (!document.body.contains(target)) {
      throw new Error("Cannot mount a component on a detached dom node");
    }
    const node = new OwlNode(this, this.Root, this.props);
    this.root = node;
    return node.mount(target);
  }

  destroy() {
    if (this.root) {
      this.root.destroy();
    }
  }
}
