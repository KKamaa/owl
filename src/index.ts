import { App } from "./app";
import { Component } from "./component";

export { App, Component };

export async function mount<T extends typeof Component>(
  C: T,
  target: HTMLElement
): Promise<InstanceType<T>> {
  const app = new App(C);
  return app.mount(target);
}

export { xml } from "./tags";
export { useState } from "./reactivity";
export { useRef } from "./refs";
export { EventBus } from "./event_bus";

export {
  onWillStart,
  onMounted,
  onBeforeUnmount,
  onWillUpdateProps,
  onBeforePatch,
  onPatched,
} from "./lifecycle_hooks";

export const __info__ = {};
