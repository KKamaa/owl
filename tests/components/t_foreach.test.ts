import { Component, mount, useState, xml } from "../../src/index";
import { makeTestFixture, nextTick, snapshotEverything } from "../helpers";

snapshotEverything();

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("list of components", () => {
  test("simple list", async () => {
    class Child extends Component {
      static template = xml`<span><t t-esc="props.value"/></span>`;
    }

    class Parent extends Component {
      static template = xml`
            <t t-foreach="state.elems" t-as="elem" t-key="elem.id">
                <Child value="elem.value"/>
            </t>`;
      static components = { Child };

      state = useState({
        elems: [
          { id: 1, value: "a" },
          { id: 2, value: "b" },
        ],
      });
    }

    const parent = await mount(Parent, fixture);
    expect(fixture.innerHTML).toBe("<span>a</span><span>b</span>");
    parent.state.elems.push({ id: 4, value: "d" });
    await nextTick();
    expect(fixture.innerHTML).toBe("<span>a</span><span>b</span><span>d</span>");

    parent.state.elems.pop();

    await nextTick();
    expect(fixture.innerHTML).toBe("<span>a</span><span>b</span>");
  });
});
