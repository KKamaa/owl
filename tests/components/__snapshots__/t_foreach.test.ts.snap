// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list of components components in a node in a t-foreach  1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-text/></div>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].item;
    return b1;
  }
}"
`;

exports[`list of components components in a node in a t-foreach  2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><ul><owl-anchor/></ul></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild.firstChild;
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<li><owl-anchor/></li>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    const b2 = b1.children[0] = new BCollection(ctx['items'], true, false);
    ctx = Object.create(ctx);
    const k1 = b2.values, v1 = b2.collection, l1 = k1.length;
    for (let i1 = 0; i1 < l1; i1++) {
      ctx[\`item\`] = v1[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v1.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k1[i1];
      let key1 = i1;
      key1 = 'li_'+ctx['item'];
      const b3 = b2.children[i1] = new Block2();
      b3.children[0] = new BComponent(\`Child\`, {item: ctx['item']}, key + \`__2__\${key1}\`, ctx, node);
      b2.keys[i1] = key1;
    }
    return b1;
  }
}"
`;

exports[`list of components reconciliation alg works for t-foreach in t-foreach 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-text/></div>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].blip;
    return b1;
  }
}"
`;

exports[`list of components reconciliation alg works for t-foreach in t-foreach 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    const b2 = b1.children[0] = new BCollection(ctx['state'].s, true, false);
    ctx = Object.create(ctx);
    const k1 = b2.values, v1 = b2.collection, l1 = k1.length;
    for (let i1 = 0; i1 < l1; i1++) {
      ctx[\`section\`] = v1[i1];
      ctx[\`section_first\`] = i1 === 0;
      ctx[\`section_last\`] = i1 === v1.length - 1;
      ctx[\`section_index\`] = i1;
      ctx[\`section_value\`] = k1[i1];
      let key1 = ctx['section_index'];
      const b3 = b2.children[i1] = new BCollection(ctx['section'].blips, false, false);
      ctx = Object.create(ctx);
      const k2 = b3.values, v2 = b3.collection, l2 = k2.length;
      for (let i2 = 0; i2 < l2; i2++) {
        ctx[\`blip\`] = v2[i2];
        ctx[\`blip_first\`] = i2 === 0;
        ctx[\`blip_last\`] = i2 === v2.length - 1;
        ctx[\`blip_index\`] = i2;
        ctx[\`blip_value\`] = k2[i2];
        let key2 = ctx['blip_index'];
        b3.children[i2] = new BComponent(\`Child\`, {blip: ctx['blip']}, key + \`__3__\${key1}__\${key2}\`, ctx, node);
        b3.keys[i2] = key2;
      }
      ctx = ctx.__proto__;
      b2.keys[i1] = key1;
    }
    return b1;
  }
}"
`;

exports[`list of components reconciliation alg works for t-foreach in t-foreach, 2 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-text/></div>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].row+'_'+ctx['props'].col;
    return b1;
  }
}"
`;

exports[`list of components reconciliation alg works for t-foreach in t-foreach, 2 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<p><owl-anchor/></p>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  class Block3 extends BNode {
    static el = elem(\`<p><owl-anchor/></p>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    const b2 = b1.children[0] = new BCollection(ctx['state'].rows, true, false);
    ctx = Object.create(ctx);
    const k1 = b2.values, v1 = b2.collection, l1 = k1.length;
    for (let i1 = 0; i1 < l1; i1++) {
      ctx[\`row\`] = v1[i1];
      ctx[\`row_first\`] = i1 === 0;
      ctx[\`row_last\`] = i1 === v1.length - 1;
      ctx[\`row_index\`] = i1;
      ctx[\`row_value\`] = k1[i1];
      let key1 = ctx['row'];
      const b3 = b2.children[i1] = new Block2();
      const b4 = b3.children[0] = new BCollection(ctx['state'].cols, true, false);
      ctx = Object.create(ctx);
      const k2 = b4.values, v2 = b4.collection, l2 = k2.length;
      for (let i2 = 0; i2 < l2; i2++) {
        ctx[\`col\`] = v2[i2];
        ctx[\`col_first\`] = i2 === 0;
        ctx[\`col_last\`] = i2 === v2.length - 1;
        ctx[\`col_index\`] = i2;
        ctx[\`col_value\`] = k2[i2];
        let key2 = ctx['col'];
        const b5 = b4.children[i2] = new Block3();
        b5.children[0] = new BComponent(\`Child\`, {row: ctx['row'],col: ctx['col']}, key + \`__3__\${key1}__\${key2}\`, ctx, node);
        b4.keys[i2] = key2;
      }
      ctx = ctx.__proto__;
      b2.keys[i1] = key1;
    }
    return b1;
  }
}"
`;

exports[`list of components simple list 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].value;
    return b1;
  }
}"
`;

exports[`list of components simple list 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new BCollection(ctx['state'].elems, false, false);
    ctx = Object.create(ctx);
    const k1 = b1.values, v1 = b1.collection, l1 = k1.length;
    for (let i1 = 0; i1 < l1; i1++) {
      ctx[\`elem\`] = v1[i1];
      ctx[\`elem_first\`] = i1 === 0;
      ctx[\`elem_last\`] = i1 === v1.length - 1;
      ctx[\`elem_index\`] = i1;
      ctx[\`elem_value\`] = k1[i1];
      let key1 = ctx['elem'].id;
      b1.children[i1] = new BComponent(\`Child\`, {value: ctx['elem'].value}, key + \`__2__\${key1}\`, ctx, node);
      b1.keys[i1] = key1;
    }
    return b1;
  }
}"
`;
